src/Collections/Sequences/NatSeq.dfy

```diff
@@ -0,0 +1,602 @@
+// RUN: %dafny /compile:0 /noNLarith "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+/*******************************************************************************
+*  Original: Copyright (c) 2020 Secure Foundations Lab
+*  SPDX-License-Identifier: MIT
+*  
+*  Modifications and Extensions: Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT 
+*******************************************************************************/
+
+/* The first element of a sequence is the least significant word; the last
+element is the most significant word. */
+
+include "../../NonlinearArithmetic/DivMod.dfy"
+include "../../NonlinearArithmetic/Mul.dfy"
+include "../../NonlinearArithmetic/Power.dfy"
+include "Seq.dfy"
+
+abstract module NatSeq {
+
+  import opened DivMod
+  import opened Mul
+  import opened Power
+  import opened Seq
+
+  /* Upper bound of a word */
+  function method BOUND(): nat
+    ensures BOUND() > 1
```

Comment:
This does seem cleaner, but we'd like to initialize `BOUND` in the refining modules

Reply:
I'm getting an error from this small example:
```
module Test {
  type LessThan10 = x | 0 <= x < 10
  const C: LessThan10
}

module Test' refines Test {
  const C: LessThan10 := 5
}
```
```
Process terminated. Assertion failed.
   at Microsoft.Dafny.Resolver.ResolveTopLevelDecls_Core(List`1 declarations, Graph`1 datatypeDependencies, Graph`1 codatatypeDependencies, Boolean isAnExport) in /private/tmp/dafny-20210205-34509-rqus4y/dafny-3.0.0/Source/Dafny/Resolver.cs:line 2644
   at Microsoft.Dafny.Resolver.ResolveModuleDefinition(ModuleDefinition m, ModuleSignature sig, Boolean isAnExport) in /private/tmp/dafny-20210205-34509-rqus4y/dafny-3.0.0/Source/Dafny/Resolver.cs:line 1002
...
```
Am I missing something?

Reply:
@RustanLeino Did you have thoughts on this issue?  I think this may be the last item holding up this PR.  If function vs. const is a minor concern, perhaps we can merge this PR and then address the bug above and consider converting to a const in a future PR.

Reply:
I think that is the right path forward. It's unfortunate since fixing this later will be a breaking change for any refining users, but worth doing while we're still reserving the right to make such changes in these libraries.

I cut https://github.com/dafny-lang/dafny/issues/1481 for the underlying issue and https://github.com/dafny-lang/libraries/issues/21 to track the improvement once that is addressed.